mlflow/utils/rest_utils.py

```diff
@@ -43,10 +44,22 @@ def http_request(host_creds, endpoint, retries=3, retry_interval=3, **kwargs):
 
     verify = not host_creds.ignore_tls_verification
 
+    def request_with_ratelimit_retries(max_rate_limit_interval, **kwargs):
+        response = requests.request(**kwargs)
+        time_left = max_rate_limit_interval
+        sleep = 1
+        while response.status_code == 429 and time_left > 0:
+            time.sleep(sleep)
+            time_left -= sleep
+            response = requests.request(**kwargs)
+            sleep = min(time_left, sleep*2)  # sleep for 1, 2, 4, ... seconds;
+        return response
+
     cleaned_hostname = strip_suffix(hostname, '/')
     url = "%s%s" % (cleaned_hostname, endpoint)
     for i in range(retries):
-        response = requests.request(url=url, headers=headers, verify=verify, **kwargs)
+        response = request_with_ratelimit_retries(max_rate_limit_interval,
```

Comment:
Just to make sure, in the absolute worst case I guess we'll retry `retries` * `num_retries_in_ request_with_ratelimit_retries` times, e.g. if we have a sequence of responses like `(429, 429, 503, 429 429, 503, 429, 429, 503)`..., but we expect that to be rare since rate-limiting is usually done by the API frontend

Reply:
Yes, exactly. So as long as the rate limiting mechanism works as expected the worst case latency should be num_retries*max_rate_limit_interval (which is equivalent to hitting api timeouts in Databricks).